src/lib/async.pl

```diff
@@ -0,0 +1,368 @@
+  
+server_async :-
+    server_options(connection_timeout_ms(Timeout)),
+    wait_for_connection(Con, Timeout),
+    await(handle_connection_async(Con)),
+    await(server_async).
+
+
+run :-
+    async_event_loop([server_async]).
+
+```
+the server and the connections now have the ability to operate asyncronously.
+
+You could define `handle_connection/1` as `handle_connection_async/1`:
+
+```prolog
+
+handle_connection_async(connection(Con)) :-
+    connection_options(connection_timeout_ms(Timeout)),
+    con_stream(Con,Stream),
+    connection_data(Data),
+    await(handle_connection_async(Data,Con,Stream,Timeout)).
+
+handle_connection_async(partial(Data),Con,Stream,Timeout) :-
+    listen_for_data(Stream,Data,NewData,Timeout),
+    report_progress(Data), % output to console?
+    await(handle_connection_async(NewData,Con,Stream,Timeout)).
+
+handle_connection_async(complete(Data),Con,_Stream,_Timeout) :-
+    await(handle_response_async(Data,Con)).
+
+```
+
+Note that there is no requirement to suffix the predicates with `_async`, but it maybe be helpful to distinguish the intent of the predicate.
+
+
+### Examples Usage
+
+```prolog
+
+
+write_async(Message) :-
+        write(Message),
+        nl.
+
+writem_async([Message|Messages]) :-
+        await(write_async(Message)),
+        await(writem_async(Messages)).
+writem_async([]).
+
+
+?- async_event_loop([writem_async([hello,world]), writem_async([cruel])]).
+   %@ hello
+   %@ cruel
+   %@ world
+   %@    true.
+```
+
+
+
+```prolog
+
+:- use_module(library(async)).
+:- use_module(library(random)).
+:- use_module(library(clpz)).
+
+countable_async(N,Max) :-
+        if_(clpz_t(N #< Max),
+            (   await(write_async(N)),
+                N1#=N+1,
+                await(countable_async(N1,Max))
+            ),
+            true
+           ).
+
+random_countable_async :-
+        await(random_integer(1,10,R)),
+        await(countable_async(0,R)).
+
+   ?- async_event_loop([random_countable_async,countable_async(0,3)]).
+   %@ 0
+   %@ 0
+   %@ 1
+   %@ 1
+   %@ 2
+   %@ 2
+   %@ 3
+   %@ 4
+   %@    true.
+
+   ?- async_event_loop([writem_async([a,b,c]),countable_async(0,3)]).
+   %@ a
+   %@ 0
+   %@ b
+   %@ 1
+   %@ c
+   %@ 2
+   %@    true.
+```
+
+## Communicating with external processes in Linux/MacOS
+
+In this example, we create a pool of asyncronous tasks that represent some amount of work being done and then waiting for them to report completion and how long they took to complete their task. Because we do this work asyncronously, completing all tasks only takes as long as the longest task (if the number of tasks is less than the number of available processors).
+
+In addition, the results are reported "as completed".
+
+```prolog
+
+:- use_module(library(async)).
+:- use_module(library(process)).
+:- use_module(library(charsio)).
+:- use_module(library(format)).
+:- use_module(library(dcgs)).
+:- use_module(library(lists)).
+:- use_module(library(time)).
+
+
+main :-
```

Comment:
Also here and the following cases: I have made good experiences with the name **`run/0`**.